### PR TITLE
feat: add retry-after header support

### PR DESCRIPTION
**Pull request by nicdard**
## Description

- Extend retries to 429 by default
- Support parsing of retry-after header and use the time specified there when available instead of the exponential backoff strategy
- Add tests to cover the new features

## Motivation and Context

This feature is used to better handle retries.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by nicdard** [comment]
What's the normal workflow? Should I rebase on top of new master or simply update the branch with merge?

File: src/HttpExponentialBackoff.js [hunk 1]
```diff
@@ -127,6 +127,25 @@ class HttpExponentialBackoff {
...
+   *  - return the value of Retry-After header, if present
+   *  - implements exponential backoff otherwise
+   *
+   * @param {number} initialDelayInMillis The multiplying factor and the initial delay. Eg. 100 would mean the retries would be spaced at 100, 200, 400, .. ms
+   * @returns {Function} retryDelayFunction {function(*=, *, *): number}
+   * @private
+   */
+  __getRetryDelayWithRetryAfterHeader (initialDelayInMillis) {
+    return (attempt, error, response) => {
+      const retryAfter = response.headers.get('Retry-After')
```
Comment (sangeetha5491) [inline src/HttpExponentialBackoff.js:142]: we need to ensure it is case-insensitive.

File: test/HttpExponentialBackoff.test.js [hunk 2]
```diff
@@ -41,6 +42,12 @@ function __testRetryOnHelper (retries, low = 499, high = 600) {
  */
 function __testRetryDelayHelper (initialDelay) {
   return jest.fn().mockImplementation(function (attempt, error, response) {
+    const retryAfter = response.headers.get('Retry-After')
+    if (retryAfter != null) {
+      const a = parseRetryAfterHeader(retryAfter)
+      console.log(a)
```
Comment (sangeetha5491) [inline test/HttpExponentialBackoff.test.js:48]: we can remove this line

File: src/HttpExponentialBackoff.js [hunk 3]
```diff
@@ -127,6 +127,25 @@ class HttpExponentialBackoff {
...
+      if (retryAfter != null) {
+        return parseRetryAfterHeader(retryAfter)
```
Comment (sangeetha5491) [inline src/HttpExponentialBackoff.js:144]: we can add a logger.debug here similar to https://github.com/adobe/aio-lib-core-networking/pull/22/commits/94b1d84e38306b6d263d4f06277a832c478efd6f#diff-d612dfd0110171e13c29ac011a0ca33d478c896fabba06a00822d8bc131f9753R125

File: src/utils.js [hunk 4]
```diff
@@ -82,7 +82,23 @@ function urlToHttpOptions (aUrl) {
...
+ *
+ * @param {string} header Retry-After header value
+ * @returns {number} Number of milliseconds to sleep until the next call to getEventsFromJournal
+ */
+function parseRetryAfterHeader (header) {
+  if (header.match(/^[0-9]+$/)) {
+    return parseInt(header, 10) * 1000
+  } else {
+    return Math.max(0, Date.parse(header) - Date.now())
+  }
```
Comment (sangeetha5491) [inline src/utils.js:97]: we can add a debug level logger here.

File: test/HttpExponentialBackoff.test.js (hunk 2 above)
Comment (nicdard) [reply test/HttpExponentialBackoff.test.js:48]: sure

File: src/HttpExponentialBackoff.js (hunk 1 above)
Comment (nicdard) [reply src/HttpExponentialBackoff.js:142]: I thought it was already as it is part of a fetch response, I was looking here: https://developer.mozilla.org/en-US/docs/Web/API/Headers/get

File: src/utils.js (hunk 4 above)
Comment (nicdard) [reply src/utils.js:98]: done

File: src/utils.js [hunk 5]
```diff
@@ -90,6 +90,7 @@ function urlToHttpOptions (aUrl) {
  * @returns {number} Number of milliseconds to sleep until the next call to getEventsFromJournal
  */
 function parseRetryAfterHeader (header) {
+  logger.debug('Parsing Retry-After header')
```
Comment (sangeetha5491) [inline src/utils.js:93]: i meant debug log to log the header value, and the returned time in milli seconds so we know what the calculated time was after which the request would be retried again. Like: https://github.com/adobe/aio-lib-core-networking/blob/master/src/HttpExponentialBackoff.js#L125

Comment (nicdard) [reply src/utils.js:93]: Oh I see, it makes more sense yeah good idea!

**Comment by sangeetha5491** [comment]
The changes look good . Will also wait for @shazron to approve

File: test/utils.test.js [hunk 6]
```diff
@@ -125,3 +126,22 @@ describe('createFetch', () => {
...
+    expect(parseRetryAfterHeader(header)).toEqual(1000)
+    expect(spy).toHaveBeenCalled()
+  })
+  test('date retry-after header older than now', () => {
+    const spy = jest.spyOn(global.Date, 'now').mockImplementation(() => new Date('Tue, 14 Feb 2023 00:00:00 GMT'))
+    const header = 'Mon, 13 Feb 2023 23:59:59 GMT'
+    expect(parseRetryAfterHeader(header)).toEqual(0)
+    expect(spy).toHaveBeenCalled()
+  })
+})
```
Comment (shazron) [inline test/utils.test.js:147]: There should be a test for the header value being a non-number and non Date. Does that return 0?

Comment (shazron) [reply test/utils.test.js:147]: If the date is older, it returns zero. (similarly my question above for malformed data, does it return 0).
Returning zero for retryDelay will cause an exception in fetch-retry@3.x: https://github.com/jonbern/fetch-retry/blob/b583df38dd3811e8258e6cd246e93c52c1f44dbb/index.js#LL19C77-L19C77

I would say if:
1. the retry-after date is older than the current date
2. the retry-after date is malformed

--> this situation is as if there is no `Retry-After` header, logically in your code. see `if (retryAfter != null) {` in `HttpExponentialBackoff.js`

**Review by shazron — CHANGES_REQUESTED** [review]

File: test/utils.test.js (hunk 6 above)
Comment (nicdard) [reply test/utils.test.js:163]: Thanks for the review! I added the required changes

**Review by shazron — APPROVED** [review]

**Pull request merged**